mmdet/models/roi_heads/standard_roi_head_with_text.py

```diff
@@ -0,0 +1,256 @@
+        if self.with_mask:
+            mask_results = self._mask_forward_train(x, sampling_results,
+                                                    bbox_results['bbox_feats'],
+                                                    gt_masks, img_metas)
+            # TODO: Support empty tensor input. #2280
+            if mask_results['loss_mask'] is not None:
+                losses.update(mask_results['loss_mask'])
+
+
+        text_results = self._text_forward_train(x, text_sampling_results, gt_masks, gt_texts, img_metas)
+        if text_results['loss_text'] is not None:
+            losses.update(text_results)
+
+        return losses
+
+    def _text_forward(self, x, rois=None, pos_inds=None, bbox_feats=None, matched_gt_texts=None):
+        assert ((rois is not None) ^
+                (pos_inds is not None and bbox_feats is not None))
+        if rois is not None:
+            text_feats = self.text_roi_extractor(
+                x[:self.text_roi_extractor.num_inputs], rois)
+            if self.with_shared_head:
+                text_feats = self.shared_head(text_feats)
+        else:
+            assert bbox_feats is not None
+            text_feats = bbox_feats[pos_inds]
+
+        text_results = self.text_head.forward(text_feats, matched_gt_texts)
+        if self.training:
+            return dict(loss_text=text_results)
+        else:
+            return dict(text_results=text_results)
+
+    def _text_forward_train(self, x, sampling_results, gt_masks, gt_texts,
+                            img_metas):
+        if not self.share_roi_extractor:
+            pos_rois = bbox2roi([res.pos_bboxes for res in sampling_results])
+            matched_gt_texts = []
+            for text, res in zip(gt_texts, sampling_results):
+                assigned_gt_indices = list(res.pos_assigned_gt_inds.cpu().numpy())
+                matched_texts = text[assigned_gt_indices]
+                assert len(matched_texts) == len(assigned_gt_indices)
+
+                matched_gt_texts.extend(matched_texts)
+            if pos_rois.shape[0] == 0:
+                return dict(loss_mask=None)
+
+            areas = (pos_rois[:, 3] - pos_rois[:, 1]) * (pos_rois[:, 4] - pos_rois[:, 2])
+            areas = areas.detach().cpu().numpy().reshape(-1)
+            # since EOS symbol added to text, subtract it
+            text_lenths = np.array([max(len(text) - 1, 1) for text in matched_gt_texts])
+
+            area_per_symbol = areas / text_lenths
+
+            matched_gt_texts = [text if aps >= self.area_per_symbol_thr else [] for text, aps in zip(matched_gt_texts, area_per_symbol)]
+
+            text_results = self._text_forward(x, pos_rois, matched_gt_texts=matched_gt_texts)
+        else:
+            raise NotImplementedError()
+
+        return text_results
+
+    def simple_test_text(self,
+                         x,
+                         img_metas,
+                         det_bboxes,
+                         det_masks,
```

Comment:
Masks are not used.

Reply:
If you don't mind I would keep it here even it is not used for now, actually it should be used for text features masking, it is highly likely that it will be used in next release